### ISSUE TITLE
Make it easier to use RuboCop with Prism

### ISSUE THREAD

**Issue by bbatsov** [rubocop/rubocop#12600]
I've noticed that these days the new super fast Prism parser has a `parser` backend that seems to work with RuboCop (see https://github.com/kddnewton/parser-prism#rubocop), but requires a bit of manual work. I've also noticed one impressive benchmark that implies RuboCop runs 4 times faster with Prism!!! 

I think we should make it easier for users to use `Prism` - e.g. if it's present it should be used by default or something along those lines. (might also be some configuration option)

@rubocop/rubocop-core Any thoughts/ideas on the subject? //cc @kddnewton

**Comment by dvandersluis** [comment]
Seems like a good idea! The only strange thing for me is the `TargetRubyVersion` overriding as that's not going to be very user-friendly.

**Comment by kddnewton** [comment]
Yeah basically I have to monkey-patch a bunch of stuff to get rubocop to recognize that prism is another possible parser. This is the file that does that: https://github.com/kddnewton/parser-prism/blob/main/lib/parser/prism/rubocop.rb

The easiest solution (if you're up for it!) would be to allow a string to be passed to `TargetRubyVersion` like `"prism"` that would fall back to `require "parser/#{target_ruby_version}"`. Other than that, anything that gets me out of having to write that horrible monkey-patch would be greatly appreciated.

In the very, very long term I would love to have a conversation with you all about the feasibility of making it the main parser. Trust me I know how much of a massive overhaul this would be, and I recognize the amount of work that would go into it. I just wanted to voice that I would be interested in exploring that kind of effort, if you were at all open to it. One small way to explore this would be trying to get just _one_ rubocop cop running directly on the prism parser. I was told 3 years ago when I pitched prism that it would take decades and probably never happen, so you never know!

**Comment by bbatsov** [comment]
> In the very, very long term I would love to have a conversation with you all about the feasibility of making it the main parser. Trust me I know how much of a massive overhaul this would be, and I recognize the amount of work that would go into it. I just wanted to voice that I would be interested in exploring that kind of effort, if you were at all open to it.

I'm definitely open to the idea, although this is something we'll have to consider very carefully - e.g. are there features of `parser` that are missing or behave differently in `prism`, can such a migration be done gradually, etc. I did the rewrite from `ripper` to `parser` early in RuboCop's life, so I have some idea how we can approach this, but also back then RuboCop was much smaller than today. (I was updating the cops in batches and for a while there were both cops using `ripper` and `parser`)

>  One small way to explore this would be trying to get just one rubocop cop running directly on the prism parser. I was told 3 years ago when I pitched prism that it would take decades and probably never happen, so you never know!

Yeah, we can certainly do some experiments in this direction. That's how I was evaluating the feasibility of `parser` early on, btw.

**Comment by rubyFeedback** [comment]
4x as fast sounds almost too good to be true. That would be an insane improvement indeed.

**Comment by koic** [comment]
Prism's remarkable achievement. In fact, I've been experimenting with parser-prism locally several times.

When it comes to integrating it into RuboCop, I think a medium to long-term perspective is necessary.

Here's the short to medium-term story first.

## Multi-Version Support

Currently lacking, but planned for future support, is multi-version capability. I learned about this in last year's RubyPrise 2023 award talk (congratulations on the award!).

Since RuboCop can distinguish between runtime Ruby version and the analysis Ruby version, multi-version support is important. For example, it's possible to analyze Ruby 3.3 source code with Ruby 3.0 runtime, and vice versa. This is a requirement for RuboCop.
https://docs.rubocop.org/rubocop/1.59/compatibility.html

## parser-prism

In terms of implementation, I think parser-prism is a very good adapter as it allows a smooth transition to Prism while retaining the Parser interface. I haven't dug deep, but it might not pass all of RuboCop's tests at the moment, so this will need to be addressed.

## Ruby Runtime Support

This is a note for a separate consideration for RuboCop.

Prism requires Ruby 3.0.0 or higher. Currently, RuboCop also includes Ruby 2.7 in its runtime, so if Ruby 2.7 is still running, a fallback to the Parser gem may be necessary.

Moreover, since Rails 7.1 supports Ruby 2.7, dropping Ruby 2.7 support might be premature.

## Universal Parser

This is a longer-term perspective, different from the above, as a Universal parser. Specifically, it relates to the parser generator Lrama, which replaced Bison in Ruby 3.3. Whether Lrama, which generates the Ruby parser based on parse.y, or Prism will become the Universal Parser, has not yet been decided by Matz.

With Ruby 3.4 (or Ruby 3.5?), the relationship between Lrama and Prism as a Universal Parser will likely progress further, so until then, it might be better to keep the Parser gem as the default and make Prism an alternative approach of some sort.

In any case, Ruby parser community is very active, and as a parser user, I'm very excited about these developments!

**Comment by bbatsov** [comment]
> With Ruby 3.4 (or Ruby 3.5?), the relationship between Lrama and Prism as a Universal Parser will likely progress further, so until then, it might be better to keep the Parser gem as the default and make Prism an alternative approach of some sort.

Yeah, I completely agree. I was mostly thinking of just simplifying the usage of `prism-parser` for the people who want to use it, most rebasing RuboCop around it. The performance gains are quite impressive and if this works reasonably well, I can imagine some people wanting to use it for CI, even if they are stuck just to the runtime version (after all most people work on Rails app that target a single runtime version for them that'd be OK).

**Comment by kddnewton** [comment]
@koic To address a couple of your points:

* Multi-Version Support: As you said this is planned. I would love to understand the true requirement here for rubocop. Is the actual requirement "be able to parse as any version of Ruby" or could that be amended to "be able to parse as the most recent version of Ruby and/or the current version of Ruby running in this process"? If it's truly "be able to parse as any version of Ruby" I wonder if we could solve this at the library level and not the parser. Because syntax is never removed, we could potentially address this kind of thing by raising an error when more recent syntax than is allowed is found. Overall, I would love more information on this in particular.
* Ruby Runtime Support: I don't imagine this will actually matter, because by the time we got this fully integrated and tested 2.7 would be EOL anyway 😅.
* Universal Parser: This shouldn't factor in. Matz has said multiple times that the parser interface going forward will be the prism AST — regardless of the technology used to create it. So the prism interface is stable and you can build on it without worrying that it's going to change out from under you.

**Comment by bbatsov** [comment]
> Multi-Version Support: As you said this is planned. I would love to understand the true requirement here for rubocop. Is the actual requirement "be able to parse as any version of Ruby" or could that be amended to "be able to parse as the most recent version of Ruby and/or the current version of Ruby running in this process"? If it's truly "be able to parse as any version of Ruby" I wonder if we could solve this at the library level and not the parser. Because syntax is never removed, we could potentially address this kind of thing by raising an error when more recent syntax than is allowed is found. Overall, I would love more information on this in particular.

The main use case is when you have a library that's targeting an older version of Ruby than the current runtime - that spares you from having to install RuboCop once per runtime and also allows users stuck to old Ruby versions to still have access to the latest RuboCop. 

Regarding syntax being accretive only - I think a couple of times some niche syntax got removed, but I'll have to double check this. But I'm certain that the meaning of some syntax changed between Ruby versions and thinks that used to be parsed in one way had to be parsed differently in later Ruby versions.

**Comment by kddnewton** [comment]
bundle[](url)

**Comment by koic** [comment]
I've obtained a direction on this. Let's keep the Parser gem as the default and make Prism an experimental option first.
Due to differences in parsers, there might be a loss of full compatibility regarding multi-versioning, but it might be possible to proceed by creating a compromise while experimenting.
Since I'm not familiar with Prism, I will likely send feedback on parser-prism (or Prism) progressively.

Including Prism as a parser option could enhance developer experience, I believe!

**Comment by kddnewton** [comment]
I've just released v0.22.0, which adds support for Ruby 2.7 (running on 2.7, not parsing _as_ 2.7). Hopefully that should make adoption easier.

w/r/t multi-versioning, we have an option now to pass in `version:` to `Prism.parse` which should help with this. I plan on creating more parsers so that you can have `Prism::Translation::Parser33` and `Prism::Translation::Parser34` and so on. We'll need a way of triggering that, so maybe the rubocop config shouldn't trigger prism based on version but instead a different option? That way you could keep `TargetRubyVersion` in place.

**Comment by koic** [comment]
I was just thinking about something similar regarding the multi-version support configuration for both existing Parser gem and Prism.

The multi-version support API of Prism (i.e., `Prism::Translation::Parser33` and `Prism::Translation::Parser34`) is great news and seems ideal for RuboCop!

This would allow for the following configuration in .rubocop.yml for parsing Ruby 3.3 with Prism:

```yaml
TargetRubyVersion: 3.3
ParserEngine: prism
```

`ParserEngine` accepts `parser` (default) and `prism` as values. It can be omitted for `ParserEngine: parser`, maintaining compatibility with traditional RuboCop. To explicitly note the oldest RuboCop support (i.e., `Parser::Ruby27`), it would look like this:

```yaml
TargetRubyVersion: 2.7
ParserEngine: parser
```

This seems to be the simplest for users at the moment.

My PoC with `Prism::Translation::Parser` appears to be working well so far. I'll think more about the design, but if there seem to be no issues, I will probably be able to open a PR soon.

@kddnewton Do you have a Prism release plan, which contains `Prism::Translation::Parser33`, `Prism::Translation::Parser34` and so on?

**Comment by kddnewton** [comment]
I have an issue open for it here: https://github.com/ruby/prism/issues/2356. But haven't gotten to it yet.

**Comment by koic** [comment]
Thank you for your sharing information. I've opened https://github.com/ruby/prism/pull/2419 to support multi-versioning for `Prism::Translation::Parser`. This should at least resolve the issue of being locked into parsing with the current Ruby 3.4.
